My Jetpack: Add red bubble and notice when paid plan needs plugin install or activation.

elliottprogrammer opened the pull request:
This PR adds a red-bubble and notice-banner when the site detects that a paid plan provides features to a plugin (or multiple plugins) that are not installed and/or activated. The notice recommends that the plugin(s) should be installed and/or activated and provides a CTA to install and activate all in one click.
To summarize, basically it adds a red-bubble and notice for missing plugins that are needed for purchased products to work. See screenshot:

![Screen Shot 2025-01-16 at 10 48 54](https://github.com/user-attachments/assets/21d18388-b76c-487d-9ce1-48369f760c48)

<!--- Provide a general summary of your changes in the Title above -->

<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes: https://github.com/Automattic/jetpack-roadmap/issues/1393
Fixes: https://github.com/Automattic/jetpack/issues/35593

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

Project Thread - pbNhbs-bJN-p2
Github Task - https://github.com/Automattic/jetpack-roadmap/issues/1393

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* It's best to test this PR with a Jurassic Ninja site (using the Jetpack Beta plugin), as opposed to a local dev site (because a local dev site already has all Jetpack plugins installed via the nature of the monorepo).
* Connect Jetpack (both site and user).
* Purchase the Jetpack Complete plan for the site.
* In My Jetpack, you should see a red-bubble and notice, stating Some plugins need installation. (See screenshot above in the description ⤴️)
* Verify all the links in the notice. The "Jetpack Complete paid subscription" link should point to the Calypso green purchases Complete subscription page. The Jetpack Boost and Jetpack CRM plugin links should point to their respective plugins page on wordpress.org
* Click the "install and activate plugins in one click" CTA.  The CTA loading text should say, "Installing and activating plugins...".
* After a few moments (after the plugins are installed and activated) the banner should disappear, and a global notice (toast/snackbar popup) in the top right corner saying "Plugin(s) installed successfully!". 
* Verify that the Jetpack Boost and CRM plugins are showing as active in My Jetpack and confirm they have been installed and activated.
* If desired, you can try other paid plan with a missing (or un-activated) plugin(s) combinations.  For example, with only a standalone plugin installed, (the main Jetpack plugin Not installed), purchase a Jetpack AI plan.  Verify a red bubble and notice appear stating that the Jetpack plugin should be installed in order for Jetpack AI to work properly.
* The notice will also appear if required plugins are installed but merely not activated, or a combination of both (not installed OR installed but not activated), so you can test Not _activated_ plugins also, if you want.

elliottprogrammer committed:
Add redbubble and notice when paid plan is missing plugin(s).

elliottprogrammer committed:
changelog

elliottprogrammer committed:
Handle undefined vars usePaidPlanNeedsPluginInstallActivationNotice hook.

elliottprogrammer committed:
Add activate-multiple-plugins endpoint & logic.

elliottprogrammer committed:
Bug fix, missing parens on implode function call.

elliottprogrammer committed:
Allow products endpoint to retrieve selected products only.

elliottprogrammer committed:
Remove notice after required plugins are installed/activated.

elliottprogrammer committed:
Delay last-backup-failed red-bubble until 30 min after purchase. Also fix $not_shown_products.

elliottprogrammer committed:
Refactor how we derive shown_products & not_shown_products.

CodeyGuyDylan commented on projects/packages/my-jetpack/src/class-rest-products.php:
```diff
@@ -217,6 +289,40 @@ public static function activate_product( $request ) {
+	public static function activate_multiple_products( $request ) {
+		$products_array = $request->get_param( 'products' );
+
+		foreach ( $products_array as $product_slug ) {
```
I have some thoughts on the notion that we need to create separate plugin endpoints for activating multiple products like this 🤔 

If we had a function to bulk activate or install plugins without requiring multiple actions, I would understand it. However, it looks like here we are essentially doing the same thing as `activate_product` with the only difference being we are looping through an array of product slugs instead of just handling one product slug. 

I could be missing something that makes my approach not the best way to go, but I think it may be better to alter the `activate_product` endpoint to optionally accept multiple products. It may require a re-naming to just `activate_products` but I think it would simplify the changes on the backend by a lot, and then on the front-end we wouldn't need the new hooks that handle multiple products that and calls a different endpoint. Let me know what you think of this, hopefully that could make this PR much smaller and simpler 😄

elliottprogrammer replied:
Thanks @CodeyGuyDylan ! Yeah I agree it would be ideal to have a shared endpoint for activating either 1 product or multiple products, but the way the My Jetpack REST endpoints are currently set up are:

_Activate endpoint:_
Path: `/my-jetpack/v1/site/products/{PRODUCT_SLUG}`
If it's a GET request: get_product()
**If it's a POST request: activate_product()**
if it's a DELETE request: deactivate_product()

Install endpoint:
Path: `/my-jetpack/v1/site/products/{PRODUCT_SLUG}/install-standalone`

So as you can see, the the endpoint route/path takes in 1 product_slug as part of the endpoint path.
Therefore I don't see how we can use that endpoint for doing anything with _multiple_ products because we can't put multiple product slugs into the endpoint route/path, unless I'm missing something, let me know.
So that's why I had to create new endpoints. I'm not sure how else I could do it.

CodeyGuyDylan reviewed (changes requested):
I am not sure why, but when running through the testing instructions with 0 standalone plugins installed, I am only seeing "CRM" as a missing plugin, not Boost
![image](https://github.com/user-attachments/assets/93cf1264-a777-4b6b-834b-bceb505c32e0)

After clicking the CTA, just as described, the banner disappears after installing the plugins (FYI it also only installed CRM, not Boost). I think this could be jarring for users. Like ".... okay? did it work?" I think it would be much better for the user experience if we had a success message after letting them know the plugins were installed. I see that this PR is pretty large already, if you'd like to handle that separately I think that'd be okay

I found a bug during testing, and I can't really tell what it causing it from the browser testing site as it seems to be PHP related. I get a critical error like this:
![image](https://github.com/user-attachments/assets/e806ab37-ffcf-4349-852f-0dfe8c0e96f7)

Steps to reproduce:
1. Uninstall main Jetpack plugin
2. Install standalone Boost plugin
3. With the beta tester, checkout this branch for the Boost plugin
4. Go to My Jetpack

I tried these steps with the stable version of boost and it's not happening so I assume it has to do with this PR. There could be other similar situations where this error takes place, I haven't yet tested all possibilities

elliottprogrammer committed:
Cleanup & add GlobalNotice after successfull activation/installation.

elliottprogrammer committed:
Merge branch 'trunk' into add/mj-red-bubble-notice-when-plugins-need-installed

elliottprogrammer committed:
Add global notice (toast) when plugins installed sucessfully.

elliottprogrammer commented:
Thanks for reviewing @CodeyGuyDylan!

> I found a bug during testing, and I can't really tell what it causing it from the browser testing site as it seems to be PHP related. I get a critical error like this:

The bug was fixed in: https://github.com/Automattic/jetpack/pull/41192

> I am not sure why, but when running through the testing instructions with 0 standalone plugins installed, I am only seeing "CRM" as a missing plugin, not Boost

Hmm, that is really strange, I'm not able to reproduce that.  The only think I can think to do is maybe try a new JN site? 🤷‍♂️ Or maybe it has something to do with the bug mentioned above (that is now fixed)?

> I think it would be much better for the user experience if we had a success message after letting them know the plugins were installed.

I added a Global Notice (a toast/snackbar popup) after the CTA action (plugins installed/activated) completes successfully. Let me know if this improves the user experience the way you had in mind.

> I have some thoughts on the notion that we need to create separate plugin endpoints for activating multiple products like this 🤔

I responded to this in the inline code comments. Let me know your thoughts. 🙂

elliottprogrammer committed:
Remove single product endpoints from mj REST_Products.

elliottprogrammer committed:
changelog

elliottprogrammer committed:
Fix static analysis possibly undeclared $product_slug.

elliottprogrammer committed:
fix php unit test.

elliottprogrammer commented:
@CodeyGuyDylan, I've refactored and cleaned up the My Jetpack Product endpoints, allowing them to act upon multiple products, and removed any endpoints that were strictly limited to acting upon only a single product, then updated the front-end to work with these changes. Also updated PHP tests.
When you have a chance, the PR is ready for review again. Thanks Dylan! 🙌

elliottprogrammer committed:
Update activate/install success toast message.

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/data/products/use-activate-plugins.ts:
```diff
@@ -0,0 +1,90 @@
+	return isPluginActive;
+};
+
+const useActivatePlugins = ( productIds: string[] ) => {
```
I think a QOL improvement for developers would be to allow either the `string[]` or `string` type, and just convert the string to an array automatically if it is not already one. That way we don't have to pass in `[ slug ]` everywhere we activate only one plugin at a time. Same with the `useInstallPlugins` and `useProducts` hooks. Do you think that would be a good idea?

elliottprogrammer replied:
Yes, thats a good idea, thanks! 👍

CodeyGuyDylan commented on projects/packages/my-jetpack/tests/php/test-products-rest.php:
```diff
@@ -167,20 +167,31 @@ public function test_get_products_not_logged() {
 
 		$response = $this->server->dispatch( $request );
 		$data     = $response->get_data();
+		print_r( $data );
```
I think we have a leftover testing line here 😄

elliottprogrammer replied:
Oops, thanks for catching that!

CodeyGuyDylan reviewed (changes requested):
This looks very good!

I did find one strange bug where the Jetpack plugin shows up twice as needing installed 
![image](https://github.com/user-attachments/assets/531966de-5993-40ca-bf13-c555682490cd)

Steps to reproduce
1. Install only the Jetpack Boost standalone plugin
2. Purchase Jetpack Complete
3. Go to My Jetpack

Not sure what's causing that, but perhaps a check to make sure every item in the list is unique would fix it? FYI the activation of the plugins does still work, so that's a plus 😄 

I left 2 other comments, one being very minor, but thank you for doing the work to update this PR. I think it's looking great and is almost ready

elliottprogrammer committed:
Filter duplicate plugins allow string or string[] useActivate & useInstall.

elliottprogrammer commented:
@CodeyGuyDylan, Thanks again for reviewing and for your feedback!
Ok, I've addressed the following:
- Filtered out displaying any duplicate plugins (in the notice/banner) that may need installed or activated.
- Allow `useActivatePlugins`, `useInstallPlugins`, and `useProducts` hooks to accept a `string` or `string[]` (array).
- Removed unintentional, left-over `print_r( $data )`.

I really appreciate all your time, attention, and effort with reviewing this PR (and all the PR's of this project!) 🙌

CodeyGuyDylan reviewed (approved):
I'm getting an issue where, if I have Complete and only Boost installed, I click the CTA and it doesn't install all the plugins it listed out in the banner. It seemed to stop after it downloaded the main Jetpack plugin. Then when I uninstall that, the banner shows me the main plugin plus the standalone's it didn't install last time. I don't think this is blocking as all the products will work (as far as I can tell) after clicking the CTA and the banner will go away.

emilyaudela commented on projects/packages/my-jetpack/_inc/data/products/use-install-plugins.ts:
```diff
@@ -0,0 +1,46 @@
+						sprintf(
+							/* translators: %s is the word "Plugin" or "Pluigns" (singular or plural). */
+							__( '%s installed successfully!', 'jetpack-my-jetpack' ),
+							products?.length === 1 ? products[ 0 ].title : __( 'Plugins', 'jetpack-my-jetpack' )
```
Would it be possible to update the plural strings to account for localization? Instead of "'%s installed successfully!", we would prefer 2 strings or a proper plural:

* plugin installed successfully!
* plugins installed successfully!

In many languages, the words will need to change depending on whether the word "plugin" is singular or plural. 

There are many instances of this in this PR. If it's not changed, some languages will not read correctly.

elliottprogrammer replied:
Hey @emilyaudela 👋!
Thanks for your feedback!  Yes! Thank you! I'll get it fixed asap. 👍🙂

elliottprogrammer replied:
Oh wait... yeah the translation comment is incorrect. The `%s` is **not** the word "Plugin" or "Plugins", but rather `%s` is the Jetpack product name, i.e.- VaultPress Backup or Boost or Social or Search or VideoPress, etc. if it's in the singular context, or `%s` is the word "Plugins" if it's in the plural context.
So yeah, anyway still, I'll fix the translation comment and make it 2 separate strings (or proper plural, whichever seems to be the best approach). 👍

elliottprogrammer replied:
I've created PR https://github.com/Automattic/jetpack/pull/41756 which separates the above translations into separate strings.
